features/192-asset-gear/index.md

```diff
@@ -0,0 +1,495 @@
+# Asset Gear
+
+## Overview
+
+The Asset Gear is a gear for uploading and retrieving user-generated asset.
+
+It supports both public and private asset.
+
+It can transform image asset on the fly via query string.
+
+The Asset Gear does not support static asset. Another gear is responsible for static asset.
+
+## Layout
+
+The Asset Gear uses a single bucket for all apps. The asset name is in the following format.
+
+```
+/<app-id>/<random-string>[.<ext>]
+```
+
+- `app-id`: The app ID.
+- `random-string`: A randomly generated string.
+- `ext`: The inferred file extension (if any) from the given content type.
+
+## HTTP API
+
+### POST /_asset/presign
+
+#### Description
+
+This endpoint returns a presigned upload request which can be used to upload an asset.
+
+This endpoint requires API Key and authenticated user.
+
+#### Request JSON Schema
+
+```json
+{
+  "type": "object",
+  "properties": {
+    "content_type": { "type": "string" },
+    "content_md5": { "type": "string" },
+    "content_length": { "type": "integer" },
+    "access": { "enum": ["public", "private"] }
+  },
+  "required": ["content_type", "content_md5", "content_length"]
+}
+```
+
+- `content_type`: The media type of the asset.
+- `content_md5`: The MD5 hash of the asset.
+- `content_length`: The size in bytes of the asset.
+- `access`: The access control of the asset. `public` is the default.
+
+##### Request Example
+
+```json
+{
+  "content_type": "image/png",
+  "content_md5": "Zum77CZrrrGRDM18nlplig==",
+  "content_length": 26419
+}
+```
+
+#### Response JSON Schema
+
+```json
+{
+  "type": "object",
+  "properties": {
+    "asset_id": { "type": "string" },
+    "url": { "type": "string" },
+    "method": { "type": "string" },
+    "headers": {
+      "type": "array",
+      "items": {
+        "type": "object",
+        "properties": {
+          "name": { "type": "string" },
+          "value": { "type": "string" }
+        }
+      }
+    }
+  },
+  "required": ["asset_id", "url", "method", "headers"]
+}
+```
+
+- `asset_id`: The asset ID. It becomes valid when the asset is uploaded.
+- `url`: The presigned URL to upload the asset.
+- `method`: The method to be used in the upload request.
+- `headers`: The headers to be included in the upload request.
+
+##### Response Example
+
+```json
+{
+  "asset_id": "myimage.png",
+  "url": "https://storage.skygearapps.com/bucket/myappid/myimage.png?...",
+  "method": "PUT",
+  "headers": [
+    { "name": "content-type", "value": "image/png" },
+    { "name": "content-md5", "value": "Zum77CZrrrGRDM18nlplig==" },
+    { "name": "content-length", "value": "26419" }
+  ]
+}
+```
+
+#### Specification
+
+1. Validate `content_type`.
+1. Let `name` be a randomly generated string.
```

Comment:
Is there any concern on not allowing customizing the asset name?

Reply:
The developer may accidentally overwrite previous asset.

Reply:
On a side note, do we not allow listing/deleting assets currently? So there is no way for developer to free up resources.

Reply:
Even we have listing and deletion, I think the developer unlikely helps us to delete unused asset.

Reply:
Not if we will charge them for it...

Would it be too limiting for use cases if the name must be randomized and not specific?

Reply:
> Is there any concern on not allowing customizing the asset name?

> Would it be too limiting for use cases if the name must be randomized and not specific?


v1 does not support user-defined asset name. Do you guys know any complaints on this limitation?

If the name is user-defined, I can think of two use cases.

## Static name

Static name like `app-logo.png`. But this use case should be solved by the static asset gear, not this gear.

## Derived name

Suppose a user has an ID of `d27a61d2-7103-4433-b67e-1374a9bc62fb`. The developer may want to name the profile image of this user as `profile-image-d27a61d2-7103-4433-b67e-1374a9bc62fb.jpg`. Not sure if this approach is useful.

Reply:
> On a side note, do we not allow listing/deleting assets currently? So there is no way for developer to free up resources.

> Not if we will charge them for it...

It is possible to add listing and deletion APIs. They should be Master Key required. I will include them in the spec. Though the developer is not going to use them unless we charge them.

Reply:
Ummm, acting like evil advocate first -- what would be the trade-off if we allow users to use custom name? More complicated becoz now we need to handle things like file override and conflict?

Reply:
## Pros and Cons of random asset name

### Pros

- Both the API and the implementation are simple. No need to deal with overwriting.
- Asset is immutable so Cache-Control header can be set trivially. Do not even require the user agent to validate the cache.

### Cons

- The developer cannot customize the asset name
- The developer cannot update the asset. They have to upload a new asset and update the reference.

## Pros and Cons of customizable name

### Pros

- Customize a part of the asset URL. Since the asset name is part of the URL, the developer can customize part of it.

### Cons

- Lose the pros of random name
- Update an asset requires two steps. The developer first uploads a new asset and then update the reference.

---

Personally I prefer random name because it is much more simpler, for both the implementation and the API. One can argue that we can allow custom name with optional Cache-Control headers to allow the developer to have fine-grained control on the caching behavior. Allowing the developer to set Cache-Control header also means they can effectively disable caching. The Asset Gear is responsible for serving the asset. Disabling caching could make the Asset Gear the bottleneck of the performance.

If custom name was allowed and conflict detection was required, the upload endpoint would be served by the Asset Gear. In current design, the upload endpoint is pointing directly to the storage server. The Asset Gear does not involve in the upload process.

Reply:
Side note

One way to deal with overwrite is to enable Object Versioning. S3 and GCE both supports Object Versioning but Azure Storage does not. [The library we are going to use does not support Object Versioning neither](https://github.com/minio/minio/pull/6494#issuecomment-534692844)

Reply:
Using random names would not allow omitting conflict detection logic: there is still a non-zero chance of ID collision. A good implementation should not cause correctness issue even if the random generated ID collided. (e.g. malicious attack of underlying randomness source?)

A use case I've seen before in mobile games: usually there will be a well-known manifest file (e.g. storage.com/mygame/manifest.json), which would contains current version and list of required data files (e.g. storage.com/mygame/enemies/v1.2.0/1021.json). This manifest file would usually served by CDN, backed by object storage (e.g. S3), since it would be retrieved very frequently. If random asset name is enforced, developer need to have an API endpoint serving the manifest file content/URL (which would not be served by CDN?), otherwise could not update the data files without publishing new app version.

Reply:
The discussion is getting long. I would like to sum it up.

**Do we need to support custom asset name?**

If the answer is no, then we need to detect accidental overwrite. We can keep the asset name uniquely in the database. A presign upload request will insert a row. Of course there will be stale rows.

If the answer is yes, then we have to

- Detect and disallow overwrite when the name is random.
- Always allow overwrite when the name is custom.
- Allow the developer to supply caching headers when the name is custom. If caching headers are not provided, a default of 15 minutes is used.

I believe both of you want custom name. If you agree with the above custom-name proposal, I will include it in the spec.

---

> A use case I've seen before in mobile games......

Your example looks like an use case of static application asset to me, which is not a goal of Asset Gear.

Reply:
> Your example looks like an use case of static application asset to me

Yes. As I understand, it would be treated as static website hosting, which is explicitly a non-goal.

Reply:
I think custom asset name is a desirable feature.

We can support two mode: developer supplied name and randomly generated name. If developer supplied name is used, it is used as-is and no duplication checking is performed. If randomly generated name is used, developer can give a prefix, and the gear would generate an asset name that is not currently in use.

Either way, the gear should not write to database when generating the upload pre-sign request: if developer want to ensure no accidental overwriting, they should check it themselves.

For caching, I think we don't need to disallow setting custom caching header when name is random.